src/transformers/models/diffllama/modeling_diffllama.py

```diff
@@ -0,0 +1,1654 @@
+            cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
+
+        key_states = repeat_kv(key_states, self.num_key_value_groups)
+        value_states = repeat_kv(value_states, self.num_key_value_groups)
+
+        causal_mask = attention_mask
+        if attention_mask is not None:
+            causal_mask = causal_mask[:, :, :, : key_states.shape[-2]]
+
+        # SDPA with memory-efficient backend is currently (torch==2.1.2) bugged with non-contiguous inputs with custom attn_mask,
+        # Reference: https://github.com/pytorch/pytorch/issues/112577.
+        if query_states.device.type == "cuda" and causal_mask is not None:
+            query_states = query_states.contiguous()
+            key_states = key_states.contiguous()
+            value_states = value_states.contiguous()
+
+        # We dispatch to SDPA's Flash Attention or Efficient kernels via this `is_causal` if statement instead of an inline conditional assignment
+        # in SDPA to support both torch.compile's dynamic shapes and full graph options. An inline conditional prevents dynamic shapes from compiling.
+        is_causal = True if causal_mask is None and q_len > 1 else False
+
+        query_states1, query_states2 = torch.chunk(query_states, 2, dim=1)
+        key_states1, key_states2 = torch.chunk(key_states, 2, dim=1)
+        value_states1, value_states2 = torch.chunk(value_states, 2, dim=1)
+
+        attn_output11 = torch.nn.functional.scaled_dot_product_attention(
+            query_states1,
+            key_states1,
+            value_states1,
+            attn_mask=causal_mask,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            is_causal=is_causal,
+        )
+        attn_output12 = torch.nn.functional.scaled_dot_product_attention(
+            query_states1,
+            key_states1,
+            value_states2,
+            attn_mask=causal_mask,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            is_causal=is_causal,
+        )
+        attn_output1 = torch.cat([attn_output11, attn_output12], dim=-1)
+
+        attn_output21 = torch.nn.functional.scaled_dot_product_attention(
+            query_states2,
+            key_states2,
+            value_states1,
+            attn_mask=causal_mask,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            is_causal=is_causal,
+        )
+        attn_output22 = torch.nn.functional.scaled_dot_product_attention(
+            query_states2,
+            key_states2,
+            value_states2,
+            attn_mask=causal_mask,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            is_causal=is_causal,
+        )
+        attn_output2 = torch.cat([attn_output21, attn_output22], dim=-1)
```

Comment:
I found we can also simplify operations as eager mode in sdpa.
```suggestion
        value_states1, value_states2 = torch.chunk(value_states, 2, dim=1)
        value_states = torch.cat([value_states1, value_states2], dim=-1)

        attn_output1 = torch.nn.functional.scaled_dot_product_attention(
            query_states1,
            key_states1,
            value_states,
            attn_mask=causal_mask,
            dropout_p=self.attention_dropout if self.training else 0.0,
            is_causal=is_causal,
        )

        attn_output2 = torch.nn.functional.scaled_dot_product_attention(
            query_states2,
            key_states2,
            value_states,
            attn_mask=causal_mask,
            dropout_p=self.attention_dropout if self.training else 0.0,
            is_causal=is_causal,
        )
```